docs/CHANGES_4.0.0.md

```diff
@@ -1,39 +1,297 @@
+### Typescript
 
-### Declare an API version on a client:
+We've migrated the driver to Typescript!
+Users can now harness the power of type hinting and intellisense in editors that support it to develop their MongoDB applications.
+Even pure JavaScript projects can benefit from the type definitions with the right linting setup.
+Along with the type hinting there's consistent and helpful docs formatting that editors should be able to display while developing.
+Recently we migrated our BSON library to TypeScript as well, this version of the driver pulls in that change.
+
+#### Community Types users (@types/mongodb)
+
+If you are a user of the community types (@types/mongodb) there will likely be compilation errors while adopting the types from our codebase.
+Unfortunately we could not achieve a one to one match in types due to the details of writing the codebase in Typescript vs definitions for the user layer API along with the breaking changes of this major version. Please let us know if there's anything that is a blocker to upgrading [on JIRA](https://jira.mongodb.org/browse/NODE).
+
+### Node.js Version
+
+We now require node 12.9 or greater for version 4 of the driver.
+If that's outside your support matrix at this time, that's okay!
+Bug fix support for our 3.x branch will not be ending until summer 2022, which has support going back as far as Node.js v4!
+
+### Cursor changes
+
+Affected classes:
+
+- `AbstractCursor`
+- `FindCursor`
+- `AggregationCursor`
+- `ChangeStreamCursor`
+  - This is the underlying cursor for `ChangeStream`
+- `ListCollectionsCursor`
+
+Our Cursor implementation has been updated to clarify what is possible before and after execution of an operation. Take this example:
 
 ```javascript
-// Declare API version "1" for the client
-client = new MongoClient(uri, { serverApi: { version: '1' } });
+const cursor = collection.find({ a: 2.3 }).skip(1);
+for await (const doc of cursor) {
+  console.log(doc);
+  fc.limit(1); // bad.
+}
+```
+
+Prior to the this release there was inconsistency surrounding how the cursor would error if a setting like limit was applied after cursor execution had begun.
+Now, an error along the lines of: `Cursor is already initialized` is thrown.
 
-cursor = client.db('database').collection('coll').find(...);
+#### ChangeStream must be used as an iterator or an event emitter
+
+You cannot use ChangeStream as iterator after using as an EventEmitter nor visa versa.
+Previously the driver would permit this kind of usage but with undefined results of which type of usage would actually receive the change documents.
+It's unlikely this kind of usage was useful but to be sure we can now warn users about the risk.
```

Comment:
```suggestion
You cannot use ChangeStream as an iterator after using as an EventEmitter nor visa versa.
Previously the driver would permit this kind of usage but it could lead to unpredictable behavior and obscure errors.
It's unlikely this kind of usage was useful but to be sure we now prevent it by throwing a clear error.
```